src/apps/orders/views.py

```diff
@@ -0,0 +1,120 @@
+
+from src.apps.basket.basket import Basket
+from .forms import DeliveryOptionForm, PaymentMethodForm
+from .models import Order, OrderProduct
+
+
+class CheckoutView(LoginRequiredMixin, View):
+    template_name = "orders/checkout.html"
+    product_list_url = reverse_lazy("product_list")
+
+    def get(self, request, *args, **kwargs):
+        basket = Basket(request)
+
+        if not basket:
+            return HttpResponseRedirect(self.product_list_url)
+
+        total_price = basket.get_total_price()
+        delivery_option_form = DeliveryOptionForm()
+        payment_method_form = PaymentMethodForm()
+
+        return render(
+            request,
+            self.template_name,
+            {
+                "basket": basket,
+                "total_price": total_price,
+                "delivery_option_form": delivery_option_form,
+                "payment_method_form": payment_method_form,
+            },
+        )
+
+    def post(self, request, *args, **kwargs):
+        basket = Basket(request)
+        total_price = basket.get_total_price()
+        delivery_option_form = DeliveryOptionForm(request.POST)
+        payment_method_form = PaymentMethodForm(request.POST)
+
+        if not basket:
+            return HttpResponseRedirect(self.product_list_url)
+
+        if delivery_option_form.is_valid() and payment_method_form.is_valid():
+            with transaction.atomic():
+                order = Order.objects.create(
+                    user_id=request.user,
+                    user_name=request.user.first_name,
+                    user_email=request.user.email,
+                )
+
+                order_products = []
+                for item in basket:
+                    product = item["product"]
+                    quantity = item["quantity"]
+                    price = item["price"]
+                    order_products.append(
+                        OrderProduct(
+                            order=order, product_id=product, quantity=quantity, product_price=price
+                        )
+                    )
+
+                OrderProduct.objects.bulk_create(order_products)
+
+                delivery_method = delivery_option_form.save(commit=False)
+                delivery_method.order = order
+                delivery_method.save()
+
+                payment_method = payment_method_form.save(commit=False)
+                payment_method.order = order
+                payment_method.save()
+
+                basket.clear()
+
+                return HttpResponseRedirect(reverse("thank_you", kwargs={"order_id": order.id}))
+
+        return render(
+            request,
+            self.template_name,
+            {
+                "basket": basket,
+                "total_price": total_price,
+                "delivery_option_form": delivery_option_form,
+                "payment_method_form": payment_method_form,
+            },
+        )
+
+
+class ThankYouView(UserPassesTestMixin, DetailView):
```

Comment:
Have you tested it with unauthorized user, I mean i have seen that authorized mixin was used together with UserPassesTestMixin but have no idea if they can be used separately

Reply:
The logic, that i was implemented is only owner of order can see thank_you/order_id/ page. And this logic works. 
Should I test smth else?

Reply:
try to access the thank_you/order_id without login and make sure we are not getting unexpected errors, thats my point

Reply:
UserPassesTestMixin will make redirect if user is not authenticated